invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalConsumableUsageInArrear.java

```diff
@@ -114,29 +113,39 @@ protected UsageInArrearDetail getToBeBilledUsageDetails(final List<RolledUpUnit>
             }
             final List<UsageConsumableInArrearTierUnitDetail> previousUsage = previousUnitsUsage.containsKey(cur.getUnitType()) ? previousUnitsUsage.get(cur.getUnitType()) : ImmutableList.<UsageConsumableInArrearTierUnitDetail>of();
 
-            usageConsumableInArrearTierUnitDetails.addAll(computeToBeBilledConsumableInArrear(cur, previousUsage, areAllBilledItemsWithDetails));
+            final List<UsageConsumableInArrearTierUnitDetail> toBeBilledConsumableInArrear = computeToBeBilledConsumableInArrear(cur, previousUsage);
+            usageConsumableInArrearTierUnitDetails.addAll(toBeBilledConsumableInArrear);
         }
         final UsageInArrearDetail toBeBilledUsageDetails = new UsageConsumableInArrearDetail(usageConsumableInArrearTierUnitDetails);
         return toBeBilledUsageDetails;
     }
 
-    private List<UsageConsumableInArrearTierUnitDetail> getUsageConsumableInArrearDetailForUnitType(final Iterable<InvoiceItem> billedItems, final String unitType) {
+    @VisibleForTesting
+    List<UsageConsumableInArrearTierUnitDetail> getBilledDetailsForUnitType(final Iterable<InvoiceItem> billedItems, final String unitType) {
 
-        final List<UsageConsumableInArrearTierUnitDetail> result = new ArrayList<UsageConsumableInArrearTierUnitDetail>();
+        // Aggregate on a per-tier level, will return a list with item per level -- for this 'unitType'
+        final Map<Integer, UsageConsumableInArrearTierUnitDetail> resultMap = new TreeMap<Integer, UsageConsumableInArrearTierUnitDetail>(Ordering.<Integer>natural());
         for (final InvoiceItem bi : billedItems) {
 
             final UsageConsumableInArrearDetail usageDetail = fromJson(bi.getItemDetails());
             for (final UsageConsumableInArrearTierUnitDetail curDetail : usageDetail.getTierDetails()) {
+
                 if (curDetail.getTierUnit().equals(unitType)) {
-                    result.add(curDetail);
+
+                    if (!resultMap.containsKey(curDetail.getTier())) {
+                        resultMap.put(curDetail.getTier(), curDetail);
+                    } else {
+                        final UsageConsumableInArrearTierUnitDetail perTierDetail = resultMap.get(curDetail.getTier());
+                        perTierDetail.updateQuantityAndAmount(perTierDetail.getQuantity() + curDetail.getQuantity());
```

Comment:
Nit: maybe `updateQuantityAndAmount(perTierDetail.getQuantity() + curDetail.getQuantity())` call could be simplified to `updateQuantityAndAmount(curDetail.getQuantity())` (i.e. the additional happens in `UsageConsumableInArrearTierUnitDetail`).

Reply:
Ok, will change the code.